libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/default_value.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{
+    ast,
+    diagnostics::DatamodelError,
+    transform::ast_to_dml::{db::ScalarType, validation_pipeline::context::Context},
+};
+use std::str::FromStr;
+
+pub(super) fn validate_default_value(
+    default_value: Option<&ast::Expression>,
+    scalar_type: Option<ScalarType>,
+    ctx: &mut Context<'_>,
+) {
+    use chrono::{DateTime, FixedOffset};
+
+    let scalar_type = if let Some(scalar_type) = scalar_type {
+        scalar_type
+    } else {
+        return;
+    };
+
+    // Scalar type specific validations.
+    match (scalar_type, default_value) {
```

Comment:
So, in general when you have so much nesting, you should maybe think about flattening it a bit. Rust is very verbose typically, so here's a diff how I would write this. It's more characters, but it's also easier to edit and copy if you need to extend the code:

```patch
diff --git a/libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/default_value.rs b/libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/default_value.rs
index 8f67aca9c7..9802b7668e 100644
--- a/libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/default_value.rs
+++ b/libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/default_value.rs
@@ -1,8 +1,12 @@
+use parser_database::ast::Expression;
+use serde_json::Value;
+
 use crate::{
     ast,
     diagnostics::DatamodelError,
     transform::ast_to_dml::{db::ScalarType, validation_pipeline::context::Context},
 };
+use chrono::{DateTime, FixedOffset};
 use std::str::FromStr;
 
 pub(super) fn validate_default_value(
@@ -10,75 +14,81 @@ pub(super) fn validate_default_value(
     scalar_type: Option<ScalarType>,
     ctx: &mut Context<'_>,
 ) {
-    use chrono::{DateTime, FixedOffset};
+    let scalar_type = match scalar_type {
+        Some(scalar_type) => scalar_type,
+        None => return,
+    };
 
-    let scalar_type = if let Some(scalar_type) = scalar_type {
-        scalar_type
-    } else {
-        return;
+    let expression = match default_value {
+        Some(default_value) => default_value,
+        None => return,
     };
 
     // Scalar type specific validations.
-    match (scalar_type, default_value) {
-        (ScalarType::Json, Some(attribute)) => {
-            if let Some((value, span)) = attribute.as_string_value() {
-                if let Err(details) = serde_json::from_str::<serde_json::Value>(value) {
-                    return ctx.push_error(DatamodelError::new_attribute_validation_error(
-                        &format!(
-                            "Parse error: \"{bad_value}\" is not a valid JSON string. ({details})",
-                            details = details,
-                            bad_value = value,
-                        ),
-                        "default",
-                        span,
-                    ));
-                }
-            }
+    match (scalar_type, expression) {
+        (ScalarType::Json, Expression::StringValue(value, span)) => {
+            let details = match serde_json::from_str::<Value>(value) {
+                Ok(_) => return,
+                Err(details) => details,
+            };
+
+            let message = format!(
+                "Parse error: \"{bad_value}\" is not a valid JSON string. ({details})",
+                details = details,
+                bad_value = value,
+            );
+
+            ctx.push_error(DatamodelError::new_attribute_validation_error(
+                &message, "default", *span,
+            ));
         }
-        (ScalarType::Bytes, Some(attribute)) => {
-            if let Some((value, span)) = attribute.as_string_value() {
-                if let Err(details) = dml::prisma_value::decode_bytes(value) {
-                    return ctx.push_error(DatamodelError::new_attribute_validation_error(
-                        &format!(
-                            "Parse error: \"{bad_value}\" is not a valid base64 string. ({details})",
-                            details = details,
-                            bad_value = value,
-                        ),
-                        "default",
-                        span,
-                    ));
-                }
-            }
+        (ScalarType::Bytes, Expression::StringValue(value, span)) => {
+            let details = match dml::prisma_value::decode_bytes(value) {
+                Ok(_) => return,
+                Err(details) => details,
+            };
+
+            let message = format!(
+                "Parse error: \"{bad_value}\" is not a valid base64 string. ({details})",
+                details = details,
+                bad_value = value,
+            );
+
+            ctx.push_error(DatamodelError::new_attribute_validation_error(
+                &message, "default", *span,
+            ));
         }
-        (ScalarType::DateTime, Some(attribute)) => {
-            if let Some((value, span)) = attribute.as_string_value() {
-                if let Err(details) = DateTime::<FixedOffset>::parse_from_rfc3339(value) {
-                    return ctx.push_error(DatamodelError::new_attribute_validation_error(
-                        &format!(
-                            "Parse error: \"{bad_value}\" is not a valid rfc3339 datetime string. ({details})",
-                            details = details,
-                            bad_value = value,
-                        ),
-                        "default",
-                        span,
-                    ));
-                }
-            }
+        (ScalarType::DateTime, Expression::StringValue(value, span)) => {
+            let details = match DateTime::<FixedOffset>::parse_from_rfc3339(value) {
+                Ok(_) => return,
+                Err(details) => details,
+            };
+
+            let message = format!(
+                "Parse error: \"{bad_value}\" is not a valid rfc3339 datetime string. ({details})",
+                details = details,
+                bad_value = value,
+            );
+
+            ctx.push_error(DatamodelError::new_attribute_validation_error(
+                &message, "default", *span,
+            ));
         }
-        (ScalarType::BigInt | ScalarType::Int, Some(attribute)) => {
-            if let Some((value, span)) = attribute.as_numeric_value() {
-                if let Err(details) = i64::from_str(value) {
-                    return ctx.push_error(DatamodelError::new_attribute_validation_error(
-                        &format!(
-                            "Parse error: \"{bad_value}\" is not a valid integer. ({details})",
-                            details = details,
-                            bad_value = value,
-                        ),
-                        "default",
-                        span,
-                    ));
-                }
-            }
+        (ScalarType::BigInt | ScalarType::Int, Expression::NumericValue(value, span)) => {
+            let details = match i64::from_str(value) {
+                Ok(_) => return,
+                Err(details) => details,
+            };
+
+            let message = format!(
+                "Parse error: \"{bad_value}\" is not a valid integer. ({details})",
+                details = details,
+                bad_value = value,
+            );
+
+            ctx.push_error(DatamodelError::new_attribute_validation_error(
+                &message, "default", *span,
+            ));
         }
         _ => (),
     }
```